File: multiprocess.py
# usage:
# python multiprocess.py [width] [height] [cores]
# if cores == 1, multiprocessing library not used

import sys
from multiprocessing import Pool

def returnNone(x):
    return None

def returnList(x):
    return [None for x in x]

if __name__ == '__main__':
    width = int(sys.argv[1])
    height = int(sys.argv[2])
    cores_str = sys.argv[3]
    no_text = cores_str[:1]
    only_text = cores_str[1:]
    cores = int(no_text)
    p = Pool(cores)
    MAXJOBS=100
    
    def maxjobs_map(f, items):
        first_i = 0
        result = []
        while first_i < len(items):
            after_last_i = first_i + MAXJOBS
            some_items = items[first_i:after_last_i]
            result += p.map(f, some_items)
            first_i += MAXJOBS
        return result
    
    def chunksize_map(f, items):
        return p.map(f, items, MAXJOBS)
    
    if cores == 1:
        MAP = map
    else:
        if "maxjobs" in cores_str:
            MAP = maxjobs_map
        elif "chunksize" in cores_str:
            MAP = chunksize_map
        else:
            MAP = p.map
    print(width, height, cores, MAP)
    if height == 1:
        base_list = list(range(width))
        MAP(returnNone, base_list)
    else:
        base_list = [list(range(height)) for x in range(width)]
        MAP(returnList, base_list)

File: multiprocess.R
works_with_R("3.2.2", data.table="1.9.6",
             "tdhock/namedCapture@6ed1257fa0d5c985931de24c3516a0f32a9323cf")

pattern <- paste0(
  "(?<trial>[0-9]+)",
  "/",
  "(?<size>[0-9]+)",
  "-",
  "(?<height>1|100)",
  "-",
  "(?<cores>[0-9])",
  "(?<maxjobs>maxjobs)?")
out.file.vec <- Sys.glob("multiprocess-data/*/*")
match.mat <- str_match_named(out.file.vec, pattern)
match.df <-
  data.frame(fun.name=ifelse(match.mat[, "cores"]==1, "map",
               ifelse(match.mat[, "maxjobs"]=="maxjobs", "maxjobs_map",
                      "Pool.map")),
             trial=match.mat[, "trial"],
             cores=as.numeric(match.mat[, "cores"]),
             pfun.name=ifelse(match.mat[, "height"]==1, "returnNone", "returnList"),
             size=as.numeric(match.mat[, "size"]))

firstValue <- function(txt){
  value <- sub(".*? ([0-9]+) .*", "\\1", txt)
  as.numeric(value)
}

increase.list <- list()
for(file.i in seq_along(out.file.vec)){
  out.file <- out.file.vec[[file.i]]
  match.row <- match.df[file.i,]
  free.lines <- readLines(out.file)
  used.lines <- grep("-/+", free.lines, value=TRUE)
  used.values <- firstValue(used.lines)
  first <- used.values[1]
  max.mem <- max(used.values)
  kilobytes <- 
    c(first.used=first,
      max.used=max.mem,
      last.used=used.values[length(used.values)],
      max.increase=max.mem-first,
      total=firstValue(free.lines[2]))
  megabytes <- as.integer(kilobytes/1024)
  increase.list[[file.i]] <- 
    data.frame(match.row,kilobytes=kilobytes[["max.increase"]])
}
multiprocess <- do.call(rbind, increase.list)

save(multiprocess, file="multiprocess.RData")
